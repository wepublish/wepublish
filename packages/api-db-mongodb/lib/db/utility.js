"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapDateFilterComparisonToMongoQueryOperatior = void 0;
const api_1 = require("@wepublish/api");
function mapDateFilterComparisonToMongoQueryOperatior(operator) {
    switch (operator) {
        case api_1.DateFilterComparison.GreaterThan:
            return '$gt';
        case api_1.DateFilterComparison.GreaterThanOrEqual:
            return '$gte';
        case api_1.DateFilterComparison.Equal:
            return '$eq';
        case api_1.DateFilterComparison.LowerThan:
            return '$lt';
        case api_1.DateFilterComparison.LowerThanOrEqual:
            return '$lte';
        default:
            throw new Error('Unknown DateFilterComparison');
    }
}
exports.mapDateFilterComparisonToMongoQueryOperatior = mapDateFilterComparisonToMongoQueryOperatior;
//# sourceMappingURL=utility.js.map