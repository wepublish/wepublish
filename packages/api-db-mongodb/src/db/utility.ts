import {DateFilterComparison} from '@wepublish/api'

export function mapDateFilterComparisonToMongoQueryOperatior(
  operator: DateFilterComparison
): string {
  switch (operator) {
    case DateFilterComparison.GreaterThan:
      return '$gt'
    case DateFilterComparison.GreaterThanOrEqual:
      return '$gte'
    case DateFilterComparison.Equal:
      return '$eq'
    case DateFilterComparison.LowerThan:
      return '$lt'
    case DateFilterComparison.LowerThanOrEqual:
      return '$lte'
    default:
      throw new Error('Unknown DateFilterComparison')
  }
}
