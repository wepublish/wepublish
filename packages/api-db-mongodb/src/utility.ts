import nanoid from 'nanoid/generate'

export const IDAlphabet = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'

export function generateID() {
  return nanoid(IDAlphabet, 16)
}

export function generateToken() {
  return nanoid(IDAlphabet, 32)
}

export function base64Encode(str: string): string {
  return Buffer.from(str).toString('base64')
}

export function base64Decode(str: string): string {
  return Buffer.from(str, 'base64').toString()
}

export function isNonNull<T>(value: T): value is NonNullable<T> {
  return value != null
}

export enum MongoErrorCode {
  DuplicateKey = 11000
}
