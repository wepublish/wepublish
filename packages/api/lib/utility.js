"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkSettingRestrictions = exports.isBoolean = exports.isString = exports.isArray = exports.isObject = exports.countRichtextChars = exports.capitalizeFirstLetter = exports.delegateToPeerSchema = exports.mapEnumsBack = exports.createProxyingIsTypeOf = exports.createProxyingResolver = exports.isSourceProxied = exports.markResultAsProxied = exports.IsProxiedSymbol = exports.base64Decode = exports.base64Encode = exports.slugify = exports.mapSubscriptionsAsCsv = exports.USER_PROPERTY_LAST_LOGIN_LINK_SEND = exports.ONE_MONTH_IN_MILLISECONDS = exports.FIFTEEN_MINUTES_IN_MILLISECONDS = exports.ONE_DAY_IN_MILLISECONDS = exports.ONE_HOUR_IN_MILLISECONDS = exports.MAX_PAYLOAD_SIZE = exports.MAX_COMMENT_LENGTH = void 0;
const formatISO_1 = __importDefault(require("date-fns/formatISO"));
const graphql_tools_1 = require("graphql-tools");
const block_1 = require("./db/block");
const error_1 = require("./error");
exports.MAX_COMMENT_LENGTH = 1000;
exports.MAX_PAYLOAD_SIZE = '1MB';
exports.ONE_HOUR_IN_MILLISECONDS = 60 * 60 * 1000;
exports.ONE_DAY_IN_MILLISECONDS = 24 * exports.ONE_HOUR_IN_MILLISECONDS;
exports.FIFTEEN_MINUTES_IN_MILLISECONDS = 900000;
exports.ONE_MONTH_IN_MILLISECONDS = 31 * exports.ONE_DAY_IN_MILLISECONDS;
exports.USER_PROPERTY_LAST_LOGIN_LINK_SEND = '_wepLastLoginLinkSentTimestamp';
function mapSubscriptionsAsCsv(subscriptions) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
    let csvStr = [
        'id',
        'firstName',
        'name',
        'preferredName',
        'email',
        'active',
        'createdAt',
        'modifiedAt',
        'company',
        'streetAddress',
        'streetAddress2',
        'zipCode',
        'city',
        'country',
        'memberPlan',
        'memberPlanID',
        'paymentPeriodicity',
        'monthlyAmount',
        'autoRenew',
        'startsAt',
        'paidUntil',
        'paymentMethod',
        'paymentMethodID',
        'deactivationDate',
        'deactivationReason'
    ].join(',') + '\n';
    for (const subscription of subscriptions) {
        const user = subscription === null || subscription === void 0 ? void 0 : subscription.user;
        const memberPlan = subscription === null || subscription === void 0 ? void 0 : subscription.memberPlan;
        const paymentMethod = subscription === null || subscription === void 0 ? void 0 : subscription.paymentMethod;
        // if (!user) continue
        csvStr +=
            [
                user === null || user === void 0 ? void 0 : user.id,
                `${sanitizeCsvContent(user === null || user === void 0 ? void 0 : user.firstName)}`,
                `${sanitizeCsvContent(user === null || user === void 0 ? void 0 : user.name)}`,
                `${sanitizeCsvContent(user === null || user === void 0 ? void 0 : user.preferredName)}`,
                `${(_a = user === null || user === void 0 ? void 0 : user.email) !== null && _a !== void 0 ? _a : ''}`,
                user === null || user === void 0 ? void 0 : user.active,
                (user === null || user === void 0 ? void 0 : user.createdAt) ? (0, formatISO_1.default)(user.createdAt, { representation: 'date' }) : '',
                (user === null || user === void 0 ? void 0 : user.modifiedAt) ? (0, formatISO_1.default)(user.modifiedAt, { representation: 'date' }) : '',
                `${sanitizeCsvContent((_b = user === null || user === void 0 ? void 0 : user.address) === null || _b === void 0 ? void 0 : _b.company)}`,
                `${sanitizeCsvContent((_c = user === null || user === void 0 ? void 0 : user.address) === null || _c === void 0 ? void 0 : _c.streetAddress)}`,
                `${sanitizeCsvContent((_d = user === null || user === void 0 ? void 0 : user.address) === null || _d === void 0 ? void 0 : _d.streetAddress2)}`,
                `${sanitizeCsvContent((_e = user === null || user === void 0 ? void 0 : user.address) === null || _e === void 0 ? void 0 : _e.zipCode)}`,
                `${sanitizeCsvContent((_f = user === null || user === void 0 ? void 0 : user.address) === null || _f === void 0 ? void 0 : _f.city)}`,
                `${sanitizeCsvContent((_g = user === null || user === void 0 ? void 0 : user.address) === null || _g === void 0 ? void 0 : _g.country)}`,
                sanitizeCsvContent(memberPlan === null || memberPlan === void 0 ? void 0 : memberPlan.name),
                (_h = subscription === null || subscription === void 0 ? void 0 : subscription.memberPlanID) !== null && _h !== void 0 ? _h : '',
                (_j = subscription === null || subscription === void 0 ? void 0 : subscription.paymentPeriodicity) !== null && _j !== void 0 ? _j : '',
                (_k = subscription === null || subscription === void 0 ? void 0 : subscription.monthlyAmount) !== null && _k !== void 0 ? _k : '',
                (_l = subscription === null || subscription === void 0 ? void 0 : subscription.autoRenew) !== null && _l !== void 0 ? _l : '',
                (subscription === null || subscription === void 0 ? void 0 : subscription.startsAt) ? (0, formatISO_1.default)(subscription.startsAt, { representation: 'date' }) : '',
                (subscription === null || subscription === void 0 ? void 0 : subscription.paidUntil)
                    ? (0, formatISO_1.default)(subscription.paidUntil, { representation: 'date' })
                    : 'no pay',
                sanitizeCsvContent(paymentMethod === null || paymentMethod === void 0 ? void 0 : paymentMethod.name),
                (_m = subscription === null || subscription === void 0 ? void 0 : subscription.paymentMethodID) !== null && _m !== void 0 ? _m : '',
                (subscription === null || subscription === void 0 ? void 0 : subscription.deactivation)
                    ? (0, formatISO_1.default)(subscription.deactivation.date, { representation: 'date' })
                    : '',
                (_p = (_o = subscription === null || subscription === void 0 ? void 0 : subscription.deactivation) === null || _o === void 0 ? void 0 : _o.reason) !== null && _p !== void 0 ? _p : ''
            ].join(',') + '\r\n';
    }
    return csvStr;
}
exports.mapSubscriptionsAsCsv = mapSubscriptionsAsCsv;
/**
 * according to rfc 4180
 * https://www.ietf.org/rfc/rfc4180.txt
 * @param input
 */
function sanitizeCsvContent(input) {
    // according rfc 4180 2.7.
    const escapeDoubleQuotes = (input || '').toString().replace(/[#"]/g, '""');
    // according rfc 4180 2.5. / 2.6.
    return `"${escapeDoubleQuotes}"`;
}
// https://gist.github.com/mathewbyrne/1280286#gistcomment-2588056
function slugify(str) {
    return str
        .toLowerCase()
        .trim()
        .replace(/[ÀÁÂÃÄÅÆĀĂĄẠẢẤẦẨẪẬẮẰẲẴẶ]/gi, 'a')
        .replace(/[ÇĆĈČ]/gi, 'c')
        .replace(/[ÐĎĐÞ]/gi, 'd')
        .replace(/[ÈÉÊËĒĔĖĘĚẸẺẼẾỀỂỄỆ]/gi, 'e')
        .replace(/[ĜĞĢǴ]/gi, 'g')
        .replace(/[ĤḦ]/gi, 'h')
        .replace(/[ÌÍÎÏĨĪĮİỈỊ]/gi, 'i')
        .replace(/[Ĵ]/gi, 'j')
        .replace(/[Ĳ]/gi, 'ij')
        .replace(/[Ķ]/gi, 'k')
        .replace(/[ĹĻĽŁ]/gi, 'l')
        .replace(/[Ḿ]/gi, 'm')
        .replace(/[ÑŃŅŇ]/gi, 'n')
        .replace(/[ÒÓÔÕÖØŌŎŐỌỎỐỒỔỖỘỚỜỞỠỢǪǬƠ]/gi, 'o')
        .replace(/[Œ]/gi, 'oe')
        .replace(/[ṕ]/gi, 'p')
        .replace(/[ŔŖŘ]/gi, 'r')
        .replace(/[ßŚŜŞŠ]/gi, 's')
        .replace(/[ŢŤ]/gi, 't')
        .replace(/[ÙÚÛÜŨŪŬŮŰŲỤỦỨỪỬỮỰƯ]/gi, 'u')
        .replace(/[ẂŴẀẄ]/gi, 'w')
        .replace(/[ẍ]/gi, 'x')
        .replace(/[ÝŶŸỲỴỶỸ]/gi, 'y')
        .replace(/[ŹŻŽ]/gi, 'z')
        .replace(/[·/_,:;\\']/gi, '-')
        .replace(/\s+/g, '-')
        .replace(/[^\w\-]+/g, '')
        .replace(/--+/g, '-')
        .replace(/^-+/, '')
        .replace(/-+$/, '');
}
exports.slugify = slugify;
function base64Encode(str) {
    return Buffer.from(str).toString('base64');
}
exports.base64Encode = base64Encode;
function base64Decode(str) {
    return Buffer.from(str, 'base64').toString();
}
exports.base64Decode = base64Decode;
exports.IsProxiedSymbol = Symbol('isProxied');
function markResultAsProxied(result) {
    if (!result)
        return null;
    for (const key in result) {
        const value = result[key];
        if (typeof value === 'object' && value != null) {
            markResultAsProxied(value);
        }
    }
    return Object.assign(result, { [exports.IsProxiedSymbol]: true });
}
exports.markResultAsProxied = markResultAsProxied;
function isSourceProxied(source) {
    return source[exports.IsProxiedSymbol] == true;
}
exports.isSourceProxied = isSourceProxied;
function createProxyingResolver(resolver) {
    return (source, args, context, info) => {
        if (isSourceProxied(source)) {
            return source[info.path.key];
        }
        return resolver(source, args, context, info);
    };
}
exports.createProxyingResolver = createProxyingResolver;
function createProxyingIsTypeOf(isTypeOf) {
    return function (source, context, info) {
        return isSourceProxied(source)
            ? source.__typename === this.name
            : isTypeOf(source, context, info);
    };
}
exports.createProxyingIsTypeOf = createProxyingIsTypeOf;
function mapEnumsBack(result) {
    if (!result)
        return null;
    for (const key in result) {
        const value = result[key];
        if (typeof value === 'object' && value !== null) {
            mapEnumsBack(value);
        }
    }
    if (result.__typename === 'ArticleTeaser' ||
        result.__typename === 'PeerArticleTeaser' ||
        result.__typename === 'PageTeaser' ||
        result.__typename === 'CustomTeaser') {
        switch (result.style) {
            case 'DEFAULT':
                return Object.assign(result, { style: block_1.TeaserStyle.Default });
            case 'LIGHT':
                return Object.assign(result, { style: block_1.TeaserStyle.Light });
            case 'TEXT':
                return Object.assign(result, { style: block_1.TeaserStyle.Text });
        }
    }
    return result;
}
exports.mapEnumsBack = mapEnumsBack;
class ResetGraphQLEnums {
    transformResult(result) {
        // FIXME: WPC-415 created
        return mapEnumsBack(result);
    }
}
async function delegateToPeerSchema(peerID, fetchAdminEndpoint, context, opts) {
    var _a;
    const schema = fetchAdminEndpoint
        ? await context.loaders.peerAdminSchema.load(peerID)
        : await context.loaders.peerSchema.load(peerID);
    if (!schema)
        return null;
    return markResultAsProxied(await (0, graphql_tools_1.delegateToSchema)(Object.assign(Object.assign({}, opts), { schema, transforms: [new ResetGraphQLEnums(), ...((_a = opts.transforms) !== null && _a !== void 0 ? _a : [])] })));
}
exports.delegateToPeerSchema = delegateToPeerSchema;
function capitalizeFirstLetter(str) {
    return str.charAt(0).toUpperCase() + str.slice(1);
}
exports.capitalizeFirstLetter = capitalizeFirstLetter;
function countRichtextChars(blocksCharLength, nodes) {
    return nodes.reduce((charLength, node) => {
        if (!node.text && !node.children)
            return charLength;
        if (node.text) {
            return charLength + node.text.length;
        }
        return countRichtextChars(charLength, node.children);
    }, blocksCharLength);
}
exports.countRichtextChars = countRichtextChars;
function isObject(unknown) {
    return typeof unknown === 'object' && unknown != null && !Array.isArray(unknown);
}
exports.isObject = isObject;
function isArray(unknown) {
    return Array.isArray(unknown);
}
exports.isArray = isArray;
function isString(unknown) {
    return typeof unknown === 'string';
}
exports.isString = isString;
function isBoolean(unknown) {
    return typeof unknown === 'boolean';
}
exports.isBoolean = isBoolean;
function checkSettingRestrictions(val, currentVal, restriction) {
    var _a, _b;
    if (!restriction) {
        return;
    }
    if (typeof val !== typeof currentVal) {
        throw new error_1.InvalidSettingValueError();
    }
    if (((_a = restriction.allowedValues) === null || _a === void 0 ? void 0 : _a.boolChoice) && typeof val !== 'boolean') {
        throw new error_1.InvalidSettingValueError();
    }
    if (typeof val === 'number') {
        if (restriction.maxValue && val > restriction.maxValue) {
            throw new error_1.InvalidSettingValueError();
        }
        if (restriction.minValue && val < restriction.minValue) {
            throw new error_1.InvalidSettingValueError();
        }
    }
    if (typeof val === 'string') {
        if (restriction.inputLength && val.length > restriction.inputLength) {
            throw new error_1.InvalidSettingValueError();
        }
        if (((_b = restriction.allowedValues) === null || _b === void 0 ? void 0 : _b.stringChoice) &&
            !restriction.allowedValues.stringChoice.includes(val)) {
            throw new error_1.InvalidSettingValueError();
        }
    }
}
exports.checkSettingRestrictions = checkSettingRestrictions;
//# sourceMappingURL=utility.js.map