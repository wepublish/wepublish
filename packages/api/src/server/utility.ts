import {sign as signJWT, verify as verifyJWT} from 'jsonwebtoken'

export enum JWTAudience {
  Refresh = 'refresh',
  Access = 'access',
  Peer = 'peer'
}

export interface RefreshToken {
  email: string
  id: string
}

export function signRefreshToken(
  id: string,
  email: string,
  tokenSecret: string,
  expiresIn: number
): string {
  return signJWT({}, tokenSecret, {
    expiresIn,
    subject: `user:${email}`,
    audience: JWTAudience.Refresh,
    jwtid: id
  })
}

export function signAccessToken(email: string, tokenSecret: string, expiresIn: number): string {
  return signJWT({}, tokenSecret, {
    expiresIn,
    subject: `user:${email}`,
    audience: JWTAudience.Access
  })
}

export function verifyRefreshToken(token: string): RefreshToken {
  const tokenData = verifyJWT(token, 'secret', {
    audience: JWTAudience.Refresh
  }) as {sub: string; jti: string}

  return {
    email: tokenData.sub.split('user:')[1],
    id: tokenData.jti
  }
}
