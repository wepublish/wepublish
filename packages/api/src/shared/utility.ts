import generateNanoID from 'nanoid/async/generate'
import generateNanoIDSync from 'nanoid/generate'

export const idAlphabet = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
export const tokenAlphabet = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_'

export function generateID() {
  return generateNanoID(idAlphabet, 10)
}

export function generateToken() {
  return generateNanoID(tokenAlphabet, 32)
}

export function generateIDSync() {
  return generateNanoIDSync(idAlphabet, 10)
}

export function generateTokenSync() {
  return generateNanoIDSync(tokenAlphabet, 32)
}
