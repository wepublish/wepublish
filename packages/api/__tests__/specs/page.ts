import {MongoDBAdapter} from '@wepublish/api-db-mongodb'
import {ApolloServerTestClient} from 'apollo-server-testing'
import {createGraphQLTestClientWithMongoDB} from '../utility'
import {
  PageInput,
  CreatePage,
  PageList,
  Page,
  UpdatePage,
  PublishPage,
  UnpublishPage,
  DeletePage
} from '../api/private'

let testClientPublic: ApolloServerTestClient
let testClientPrivate: ApolloServerTestClient
let dbAdapter: MongoDBAdapter

beforeAll(async () => {
  try {
    const setupClient = await createGraphQLTestClientWithMongoDB()
    testClientPublic = setupClient.testClientPublic
    testClientPrivate = setupClient.testClientPrivate
    dbAdapter = setupClient.dbAdapter

    console.log('public', testClientPublic)
  } catch (error) {
    console.log('Error', error)
    throw new Error('Error during test setup')
  }
})

describe('Pages', () => {
  describe('can be created/edited/published/unpublished/deleted:', () => {
    const ids: string[] = []
    beforeEach(async () => {
      const {mutate} = testClientPrivate
      const input: PageInput = {
        title: 'Testing Page',
        slug: 'testing-page',
        tags: ['one', 'two', 'three'],
        properties: [
          {key: 'private', value: 'private', public: false},
          {key: 'public', value: 'public', public: true}
        ],
        blocks: []
      }
      const res = await mutate({
        mutation: CreatePage,
        variables: {
          input: input
        }
      })
      ids.unshift(res.data?.createPage?.id)
    })

    test('can be created', async () => {
      const {mutate} = testClientPrivate
      const input: PageInput = {
        title: 'Testing Page',
        slug: 'testing-page',
        tags: ['one', 'two', 'three'],
        properties: [
          {key: 'private', value: 'private', public: false},
          {key: 'public', value: 'public', public: true}
        ],
        blocks: [
          {
            title: {
              title: 'title block title',
              lead: 'title lead'
            }
          }
        ]
      }
      const res = await mutate({
        mutation: CreatePage,
        variables: {
          input: input
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          createPage: {
            id: expect.any(String)
          }
        }
      })
      ids.unshift(res.data?.createPage?.id)
    })

    test('can be read in list', async () => {
      const {query} = testClientPrivate
      const res = await query({
        query: PageList,
        variables: {
          first: 100
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          pages: {
            nodes: Array.from({length: ids.length}, () => ({
              createdAt: expect.any(String),
              id: expect.any(String),
              modifiedAt: expect.any(String)
            })),
            pageInfo: {
              endCursor: expect.any(String),
              startCursor: expect.any(String)
            },
            totalCount: expect.any(Number)
          }
        }
      })
      expect(res.data?.pages?.totalCount).toBe(ids.length)
    })

    test('can be read by id', async () => {
      const {query} = testClientPrivate
      const res = await query({
        query: Page,
        variables: {
          id: ids[0]
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          page: {
            id: expect.any(String)
          }
        }
      })
    })

    test('can be updated', async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: UpdatePage,
        variables: {
          input: {
            title: 'Update Title',
            slug: 'update-slug',
            tags: ['update', 'tags'],
            properties: [
              {key: 'private', value: 'private', public: false},
              {key: 'public', value: 'public', public: true}
            ],
            blocks: [
              {
                title: {
                  title: 'new updated title',
                  lead: 'update lead'
                }
              }
            ]
          },
          id: ids[0]
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          updatePage: {
            id: expect.any(String)
          }
        }
      })
    })

    test('can be published', async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: PublishPage,
        variables: {
          id: ids[0],
          publishAt: '2020-11-25T23:55:35.000Z',
          publishedAt: '2020-11-25T23:55:35.000Z',
          updatedAt: '2020-11-25T23:55:35.000Z'
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          publishPage: {
            id: expect.any(String)
          }
        }
      })
    })

    test('can be unpublished', async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: UnpublishPage,
        variables: {
          id: ids[0]
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          unpublishPage: {
            id: expect.any(String)
          }
        }
      })
    })

    test('can be deleted', async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: DeletePage,
        variables: {
          id: ids[0]
        }
      })
      expect(res).toMatchSnapshot()
      ids.shift()
    })
  })
})

afterAll(async () => {
  if (dbAdapter) {
    await dbAdapter.db.dropDatabase()
    await dbAdapter.client.close()
  }
})
