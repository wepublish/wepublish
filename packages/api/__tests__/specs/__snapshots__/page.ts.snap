// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages can be created/edited/published/unpublished/deleted: can be created 1`] = `
Object {
  "data": Object {
    "createPage": Object {
      "draft": Object {
        "publishedAt": null,
        "revision": 0,
        "updatedAt": null,
      },
      "id": Any<String>,
      "latest": Object {
        "blocks": Array [
          Object {
            "__typename": "TitleBlock",
            "lead": "title lead",
            "title": "title block title",
          },
          Object {
            "__typename": "ListicleBlock",
            "items": Array [
              Object {
                "image": null,
                "richText": Array [
                  Object {
                    "children": Array [
                      Object {
                        "text": "p text rich text",
                      },
                    ],
                    "type": "paragraph",
                  },
                ],
                "title": "item title",
              },
            ],
          },
          Object {
            "__typename": "RichTextBlock",
            "richText": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "p text rich text",
                  },
                ],
                "type": "paragraph",
              },
            ],
          },
          Object {
            "__typename": "FacebookPostBlock",
            "postID": "2191372684210495",
            "userID": "WePublish.Community",
          },
          Object {
            "__typename": "FacebookVideoBlock",
            "userID": "WePublish.Community",
            "videoID": "2F1761095367238231",
          },
          Object {
            "__typename": "InstagramPostBlock",
            "postID": "B_AfAKgARJ4",
          },
          Object {
            "__typename": "TwitterTweetBlock",
            "tweetID": "1021701711172395008",
            "userID": "WePublish_media",
          },
          Object {
            "__typename": "VimeoVideoBlock",
            "videoID": "472380240",
          },
          Object {
            "__typename": "YouTubeVideoBlock",
            "videoID": "FjtGLMYgY5Y",
          },
          Object {
            "__typename": "SoundCloudTrackBlock",
            "trackID": "310511381",
          },
          Object {
            "__typename": "EmbedBlock",
            "height": null,
            "styleCustom": null,
            "title": null,
            "url": "https://www.youtube.com/embed/cPAbx5kgCJo",
            "width": null,
          },
          Object {
            "__typename": "LinkPageBreakBlock",
            "linkText": null,
            "linkURL": null,
            "text": null,
          },
          Object {
            "__typename": "TeaserGridBlock",
            "numColumns": 2,
            "teasers": Array [],
          },
          Object {
            "__typename": "QuoteBlock",
            "author": "author",
            "quote": "quote text",
          },
        ],
        "description": null,
        "image": null,
        "properties": Array [
          Object {
            "key": "private",
            "public": false,
            "value": "private",
          },
          Object {
            "key": "public",
            "public": true,
            "value": "public",
          },
        ],
        "publishedAt": null,
        "slug": "testing-page",
        "tags": Array [
          "one",
          "two",
          "three",
        ],
        "title": "Testing Page",
        "updatedAt": null,
      },
      "pending": null,
      "published": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Pages can be created/edited/published/unpublished/deleted: can be deleted 1`] = `
Object {
  "data": Object {
    "deletePage": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Pages can be created/edited/published/unpublished/deleted: can be published 1`] = `
Object {
  "data": Object {
    "publishPage": Object {
      "draft": null,
      "id": Any<String>,
      "latest": Object {
        "blocks": Array [],
        "description": null,
        "image": null,
        "properties": Array [
          Object {
            "key": "private",
            "public": false,
            "value": "private",
          },
          Object {
            "key": "public",
            "public": true,
            "value": "public",
          },
        ],
        "publishedAt": "2020-11-25T23:55:35.000Z",
        "slug": "testing-page",
        "tags": Array [
          "one",
          "two",
          "three",
        ],
        "title": "Testing Page",
        "updatedAt": "2020-11-25T23:55:35.000Z",
      },
      "pending": null,
      "published": Object {
        "publishedAt": "2020-11-25T23:55:35.000Z",
        "revision": 0,
        "updatedAt": "2020-11-25T23:55:35.000Z",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Pages can be created/edited/published/unpublished/deleted: can be read by id 1`] = `
Object {
  "data": Object {
    "page": Object {
      "id": Any<String>,
      "latest": Object {
        "blocks": Array [],
        "description": null,
        "image": null,
        "properties": Array [
          Object {
            "key": "private",
            "public": false,
            "value": "private",
          },
          Object {
            "key": "public",
            "public": true,
            "value": "public",
          },
        ],
        "publishedAt": null,
        "slug": "testing-page",
        "tags": Array [
          "one",
          "two",
          "three",
        ],
        "title": "Testing Page",
        "updatedAt": null,
      },
      "pending": null,
      "published": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Pages can be created/edited/published/unpublished/deleted: can be read in list 1`] = `
Object {
  "data": Object {
    "pages": Object {
      "nodes": Array [
        Object {
          "createdAt": Any<String>,
          "draft": Object {
            "revision": 0,
          },
          "id": Any<String>,
          "latest": Object {
            "description": null,
            "image": null,
            "publishedAt": null,
            "revision": 0,
            "title": "Testing Page",
            "updatedAt": null,
          },
          "modifiedAt": Any<String>,
          "pending": null,
          "published": null,
        },
        Object {
          "createdAt": Any<String>,
          "draft": Object {
            "revision": 0,
          },
          "id": Any<String>,
          "latest": Object {
            "description": null,
            "image": null,
            "publishedAt": null,
            "revision": 0,
            "title": "Testing Page",
            "updatedAt": null,
          },
          "modifiedAt": Any<String>,
          "pending": null,
          "published": null,
        },
        Object {
          "createdAt": Any<String>,
          "draft": Object {
            "revision": 0,
          },
          "id": Any<String>,
          "latest": Object {
            "description": null,
            "image": null,
            "publishedAt": null,
            "revision": 0,
            "title": "Testing Page",
            "updatedAt": null,
          },
          "modifiedAt": Any<String>,
          "pending": null,
          "published": null,
        },
      ],
      "pageInfo": Object {
        "endCursor": Any<String>,
        "hasNextPage": false,
        "hasPreviousPage": false,
        "startCursor": Any<String>,
      },
      "totalCount": Any<Number>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Pages can be created/edited/published/unpublished/deleted: can be unpublished 1`] = `
Object {
  "data": Object {
    "unpublishPage": Object {
      "draft": Object {
        "publishedAt": null,
        "revision": 0,
        "updatedAt": null,
      },
      "id": Any<String>,
      "latest": Object {
        "blocks": Array [],
        "description": null,
        "image": null,
        "properties": Array [
          Object {
            "key": "private",
            "public": false,
            "value": "private",
          },
          Object {
            "key": "public",
            "public": true,
            "value": "public",
          },
        ],
        "publishedAt": null,
        "slug": "testing-page",
        "tags": Array [
          "one",
          "two",
          "three",
        ],
        "title": "Testing Page",
        "updatedAt": null,
      },
      "pending": null,
      "published": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Pages can be created/edited/published/unpublished/deleted: can be updated 1`] = `
Object {
  "data": Object {
    "updatePage": Object {
      "draft": Object {
        "publishedAt": null,
        "revision": 0,
        "updatedAt": null,
      },
      "id": Any<String>,
      "latest": Object {
        "blocks": Array [
          Object {
            "__typename": "TitleBlock",
            "lead": "update lead",
            "title": "new updated title",
          },
        ],
        "description": null,
        "image": null,
        "properties": Array [
          Object {
            "key": "private",
            "public": false,
            "value": "private",
          },
          Object {
            "key": "public",
            "public": true,
            "value": "public",
          },
        ],
        "publishedAt": null,
        "slug": "update-slug",
        "tags": Array [
          "update",
          "tags",
        ],
        "title": "Update Title",
        "updatedAt": null,
      },
      "pending": null,
      "published": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Pages can be created/edited/published/unpublished/deleted: can not be published with the same slug 1`] = `
Object {
  "data": Object {
    "publishPage": null,
  },
  "errors": Any<Array>,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
